vllm/model_executor/guided_decoding.py

```diff
@@ -0,0 +1,113 @@
+import asyncio
+from collections import defaultdict
+import concurrent.futures
+from copy import copy
+from enum import Enum
+from functools import lru_cache
+from json import dumps as json_dumps
+from re import escape as regex_escape
+from typing import Union, Tuple
+from types import SimpleNamespace
+from pydantic import BaseModel
+
+from vllm.entrypoints.openai.protocol import CompletionRequest, ChatCompletionRequest
+from vllm.model_executor.guided_logits_processors import JSONLogitsProcessor, RegexLogitsProcessor
+
+
+class GuidedDecodingMode(Enum):
+    JSON = "json"
+    REGEX = "regex"
+    CHOICE = "choice"
+
+
+async def get_guided_decoding_logits_processor(
+        request: Union[CompletionRequest, ChatCompletionRequest],
+        tokenizer
+    ) -> Union[JSONLogitsProcessor, RegexLogitsProcessor]:
+    """
+    Given an OpenAI-compatible request, check for guided decoding parameters
+    and get the necessary logits processor for the given guide.
+    We cache logit processors by (guide, tokenizer), and on cache hit
+    we make a shallow copy to reuse the same underlying FSM.
+    """
+    guide, mode = _get_guide_and_mode(request)
+    if not guide:
+        return None
+    loop = asyncio.get_running_loop()
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=2) as pool:
+        result = await loop.run_in_executor(
+            pool, get_cached_logits_processor,
+            guide, tokenizer, mode
+        )
+        logits_processor = copy(result)
+        # reset logits processor's internal state
+        logits_processor.fsm_state = defaultdict(int)
```

Comment:
**nitpick** (no blocker)

Now that the logits processors are part of vLLM, we could introduce a reset_state method to them instead of changing internal fields from outside.